src/Microsoft.PowerShell.ConsoleHost/host/msh/Executor.cs

```diff
@@ -320,11 +322,346 @@ internal Collection<PSObject> ExecuteCommand(string command, out Exception excep
+                        // Check for allowed command
+                        var aliasInfo = cmdInfo as AliasInfo;
+                        string cmdName = (aliasInfo != null) ? aliasInfo.ReferencedCommand.Name : cmdInfo.Name;
+                        if (AllowedCommands.Contains(cmdName))
+                        {
+                            continue;
+                        }
+
+                        if (cmdInfo.Module == null || string.IsNullOrEmpty(cmdInfo.ModuleName))
+                        {
+                            success = false;
+                            break;
+                        }
+
+                        var privateData = cmdInfo.Module.PrivateData as System.Collections.Hashtable;
+                        if (privateData == null)
+                        {
+                            success = false;
+                            break;
+                        }
+                        if (!privateData.ContainsKey("ImplicitSessionId"))
+                        {
+                            success = false;
+                            break;
+                        }
+                        if (psSessionId == Guid.Empty)
+                        {
+                            psSessionId = new Guid(privateData["ImplicitSessionId"] as string);
+                        }
+                        else if (!psSessionId.ToString().Equals(privateData["ImplicitSessionId"] as string, StringComparison.OrdinalIgnoreCase))
+                        {
+                            success = false;
+                            break;
+                        }
+                    }
+
+                    if (success)
+                    {
+                        //
+                        // Invoke command pipeline as entire pipeline on remote session
+                        //
+
+                        // Update script to declare variables via Using keyword
+                        if (checker.ValidVariables.Count > 0)
+                        {
+                            foreach (var variableName in checker.ValidVariables)
+                            {
+                                command = command.Replace(variableName, ("Using:" + variableName), StringComparison.OrdinalIgnoreCase);
+                            }
+
+                            scriptBlock = ScriptBlock.Create(command);
+                        }
+
+                        // Retrieve the PSSession runspace in which to run the batch script on
+                        ps.Commands.Clear();
+                        ps.Commands.AddCommand("Get-PSSession").AddParameter("InstanceId", psSessionId);
+                        var psSession = ps.Invoke<System.Management.Automation.Runspaces.PSSession>().FirstOrDefault();
+                        if (psSession == null || ps.Streams.Error.Count > 0)
+                        {
+                            // Cannot continue
+                            throw new PSInvalidOperationException();
```

Comment:
Do we need a better exception message here?

Reply:
Will add.